crates/task-impls/src/quorum_proposal_recv/handlers.rs

```diff
@@ -0,0 +1,232 @@
+    },
+    events::HotShotEvent,
+    helpers::broadcast_event,
+};
+
+use super::QuorumProposalRecvTaskState;
+
+/// Broadcast the liveness proposal in the event that the parent state is not found for
+/// a given `proposal`.
+async fn send_liveness_proposal<TYPES: NodeType, I: NodeImplementation<TYPES>>(
+    proposal: &Proposal<TYPES, QuorumProposal<TYPES>>,
+    event_sender: &Sender<Arc<HotShotEvent<TYPES>>>,
+    justify_qc: &QuorumCertificate<TYPES>,
+    task_state: &mut QuorumProposalRecvTaskState<TYPES, I>,
+) -> Option<QuorumProposal<TYPES>> {
+    let view_number = proposal.data.view_number();
+    let mut consensus_write = task_state.consensus.write().await;
+
+    warn!(
+        "Proposal's parent missing from storage with commitment: {:?}",
+        justify_qc.data.leaf_commit
+    );
+    let leaf = Leaf::from_quorum_proposal(&proposal.data);
+
+    let state = Arc::new(
+        <TYPES::ValidatedState as ValidatedState<TYPES>>::from_header(&proposal.data.block_header),
+    );
+    let view = View {
+        view_inner: ViewInner::Leaf {
+            leaf: leaf.commit(),
+            state,
+            delta: None,
+        },
+    };
+
+    consensus_write.update_validated_state_map(view_number, view.clone());
+    consensus_write.update_saved_leaves(leaf.clone());
+
+    if let Err(e) = task_state
+        .storage
+        .write()
+        .await
+        .update_undecided_state(
+            consensus_write.saved_leaves().clone(),
+            consensus_write.validated_state_map().clone(),
+        )
+        .await
+    {
+        warn!("Couldn't store undecided state.  Error: {:?}", e);
+    }
+
+    let liveness_check = justify_qc.view_number() > consensus_write.locked_view();
+
+    let high_qc = consensus_write.high_qc().clone();
+    drop(consensus_write);
+
+    // Broadcast that we've updated our consensus state so that other tasks know it's safe to grab.
+    broadcast_event(
+        HotShotEvent::ValidatedStateUpdated(view_number, view).into(),
+        event_sender,
+    )
+    .await;
+    broadcast_event(
+        HotShotEvent::NewUndecidedView(leaf.clone()).into(),
+        event_sender,
+    )
+    .await;
+
+    let mut current_proposal = None;
+    if liveness_check {
+        current_proposal = Some(proposal.data.clone());
+
+        let new_view = proposal.data.view_number + 1;
+
+        // This is for the case where we form a QC but have not yet seen the previous proposal ourselves
+        let should_propose = task_state.quorum_membership.leader(new_view) == task_state.public_key
+            && high_qc.view_number == current_proposal.clone().unwrap().view_number;
+
+        if should_propose {
+            debug!(
+                "Attempting to publish proposal after voting for liveness; now in view: {}",
+                *new_view
+            );
+            broadcast_event(
+                HotShotEvent::LivenessCheckProposalRecv(proposal.data.clone()).into(),
```

Comment:
Similar to the suggestion for the function name, maybe this could be `ProposalLivenessValidated`.

Reply:
Ah yeah that's a good name. Perhaps I'll do `QuorumProposalLivenessValidated` or something like that. Though I'm not sure if that would be an disambiguation or not. I'll think about it.

Reply:
Since we send a quorum proposal, I opted for the name above.